File: haskell/strain/Strain.hs
module Strain (keep, discard) where

keep :: (a -> Bool) -> [a] -> [a]
keep p =
  foldr keep' []
  where
    keep' x xs = if p x then x:xs else xs

discard :: (a -> Bool) -> [a] -> [a]
discard = keep . (not . )

File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[T](ts: Seq[T], f: T => Boolean): Seq[T] =
    ts.foldRight(Seq.empty[T]) { case (t, ys) =>
      if (f(t)) t +: ys else ys
    }

  def discard[T](ts: Seq[T], f: T => Boolean): Seq[T] =
    keep(ts, !f(_:T))
}